### DISCUSSION TITLE
C and CXX compiler identification unknown while building using docker

### DISCUSSION THREAD

**Discussion by kshitij01042002** [trustwallet/wallet-core#3608] [Q&A]
**Describe the bug**
When attempting to build wallet-core using the provided [Docker image installation](https://developer.trustwallet.com/developer/wallet-core/developing-the-library/building#building-inside-docker-image) , the build process encounters an issue related to the C and C++ compilers.

**To Reproduce**
- Use the provided [Docker image installation](https://developer.trustwallet.com/developer/wallet-core/developing-the-library/building#building-inside-docker-image) from the wallet-core website.
- Attempt to build Trust Wallet using the provided instructions.
- Observe the error mentioned above during the build process.

**Expected behavior**
The build shoud be successfull.

**Screenshots**
![image](https://github.com/trustwallet/wallet-core/assets/91470808/2aba4792-fb58-48a6-beb2-add585e0c521)

**Additional context**

Logs in .`/wallet-core# cat build/CMakeFiles/CMakeError.log`

![Screenshot from 2023-12-18 10-35-39](https://github.com/trustwallet/wallet-core/assets/91470808/90365325-4d3f-451c-a0e6-2180be91faf5)

**Comment by satoshiotomakan** [comment]
Hi @kshitij01042002, this error is unexpected if you do not modify `Dockerfile`.
Could you please provide the exact command you use to build the Docker image?

**Comment by kshitij01042002** [comment]
Hi @satoshiotomakan, thankyou for your response. I have not modified the Dockerfile.

I have used the steps mentioned [here](https://developer.trustwallet.com/developer/wallet-core/developing-the-library/building#building-inside-docker-image).

For the exact commands:

1. `docker build docker/wallet-core --tag wallet-core-dev`
2. `docker run -i -t wallet-core-dev /bin/bash`
3. `./tools/build-and-test`

On running the command mentioned in the step 3 we get the following error.

**Comment by AnilChinchawale** [comment]
Hello @satoshiotomakan Can you share instructions that we can follow to verify test cases in the local environment. 

Thanks.

**Reply by satoshiotomakan** [reply]
Found out the issue. The required environment variables are set in the Dockerfile as `ENV` command. They are valid during the Docker build only, so when you run the docker container with `/bin/bash`, the variables are no longer available.
https://github.com/trustwallet/wallet-core/blob/b530432921d7a9709428b0162673e0ab72de1c3d/Dockerfile#L34-L35
https://github.com/trustwallet/wallet-core/blob/b530432921d7a9709428b0162673e0ab72de1c3d/Dockerfile#L40

So to run the `tools/build-and-test` command inside a docker container, you need to set the variables manually:
```bash
export CC=/usr/bin/clang-14
export CXX=/usr/bin/clang++-14
export PATH="/root/.cargo/bin:${PATH}"

./tools/build-and-test
```

However, the docker image you built with `docker build` has already compiled the static library and the test binary, so you can easily run the tests with
```
./build/tests/tests
```

**Reply by AnilChinchawale** [reply]
Thank you @satoshiotomakan , Let us try the above method quickly.

**Comment by kshitij01042002** [comment]
Hi @satoshiotomakan I tried by setting the variables inside the Docker container but the error still persist. Though I am able to run the tests inside the container using `./build/tests/tests` and build using the command `sudo docker build . --tag wallet-core-dev`.
